feat: Add Talos operator

alperencelik opened the pull request:
Talos-operator is a Kubernetes operator for managing Talos Linux clusters using Custom Resources. It offers two primary modes: "container" and "metal". The container mode allows you to bootstrap Talos Linux clusters as Kubernetes pods. The metal mode supports physical and virtual machines that have been booted with Talos Linux and are waiting in maintenance mode. 

Repo: https://github.com/alperencelik/talos-operator

Documentation: https://alperencelik.github.io/talos-operator

alperencelik commented:
Although the project is a work in progress, I believe that it has reached a stage where it's ready to share it.

rothgar commented:
This is neat. Is it cluster API based?

alperencelik commented:
Hi @rothgar, thanks for your kind words. It doesn’t use ClusterAPI or anything related to it. It’s a proprietary Kubernetes operator that consumes the Talos API. It doesn’t handle anything related to Infrastructure(**), so it requires Talos machines to be up and available in “Maintenance” mode. Essentially, it’s a bootstrapper that runs the equivalent of the talosctl commands to bootstrap a cluster.

**: For the infrastructure part, I use my [Proxmox operator](https://github.com/alperencelik/kubemox) since I primarily operate there. I’m also looking into integrating [boot-to-talos](https://github.com/cozystack/boot-to-talos), at least for converting a machine to Talos Linux.

rothgar commented:
You may also want to look into creating an [infrastructure provider](https://omni.siderolabs.com/explanation/infrastructure-providers) for proxmox. I suspect it would simplify a lot of the work you've done and is intended to create VMs and clusters automatically in these kinds of environments.

rothgar commented on README.md:
```diff
@@ -59,6 +59,7 @@ Collection of awesome talos resource from the community
 - [terraform vsphere cluster](https://github.com/ilpozzd/terraform-talos-vsphere-cluster) Deploy a Kubernetes cluster based on Talos OS in vSphere
 - [terraform vsphere vm](https://github.com/ilpozzd/terraform-talos-vsphere-vm) Deploy a Talos OS-based vSphere virtual machine in vSphere
 - [Talos Hetzner Dedicated Control](https://github.com/ErikLundJensen/thdctl) CLI to create Talos based Kubernetes clusters with Hetzner dedicated servers
+- [Talos Operator](https://github.com/alperencelik/talos-operator) A Kubernetes operator to manage Talos Linux clusters on Kubernetes
```
Does this only work with Proxmox? We should probably mention it here.

alperencelik replied:
No, it has mainly two modes here. One is called "container" which deploys Talos nodes as Kubernetes pods.

The other mode is metal and for this mode the only requirement is Talos Linux booted machine sitting at the maintenance mode. It doesn't have to be in Proxmox or any other hypervisor. It could be either physical or virtual machine.

alperencelik replied:
Hi @rothgar, do you think we should update the description, or is the existing one good?

alperencelik commented:
> You may also want to look into creating an [infrastructure provider](https://omni.siderolabs.com/explanation/infrastructure-providers) for proxmox. I suspect it would simplify a lot of the work you've done and is intended to create VMs and clusters automatically in these kinds of environments.

Thanks, I’ll definitely look into these. 

The main purpose of the operator is to replace the talosctl commands normally required to bootstrap a cluster. In my case, I had to boot machines with Talos Linux, and once the infrastructure was ready, I had to create configs, apply them to each machine individually, and perform other manual steps such as Talos upgrades and Kubernetes upgrades.

With the operator, you don’t have to deal with bundle secrets, configs, or similar tasks. It streamlines config generation based on the input fields defined in the CRD. If you have a more complex configuration that isn’t supported by the operator’s fields, you can provide your own configs via a ConfigMap, and the operator will apply them to your machines. Even then, you still avoid handling secrets or manually running CLI commands such as version upgrades or Kubernetes upgrades.

leo-baltus commented:
It seems your documentation is lacking a bit.

- ``helm repo add talos-operator https://alperencelik.github.io/talos-operator`` results in a 404
- it is unclear which node is targeted when using ``helm install talos-operator talos-operator/talos-operator``
- also it does not mention why it can do without Talos secrets. Where are they and do they get renewed? How can I get a kubectl config?

alperencelik commented:
@leo-baltus Thanks for the feedback! I am addressing the issues you mentioned regarding the Helm chart and config files.

- I just updated the helm repo link so please try the following: `helm repo add talos-operator https://alperencelik.github.io/helm-charts`

- Well, helm command by default uses your current-context, if you need to update for a specific context you can always set the `--kube-context` flag. The operator installation doesn't target anything or perform any operation. The operator will start to work as soon as you create or update Custom Resource(s),

- I just tried to cover this one in the documentation but basically, operator will generate the required secret bundle by itself (like talosctl command does) and currently it doesn't support any rotation on that secret bundle.

You can retrieve both Talosconfig and Kubeconfig from the secrets created by the controller. The naming convention as follows: [talos-cluster-name]-talosconfig &  [talos-cluster-name]-kubeconfig

leo-baltus commented:
> * `helm repo add talos-operator https://alperencelik.github.io/helm-charts`

check. works.

Regarding helm. Maybe I do not understand your use case. In my case I want to deploy on bare metal. So maybe I should not use helm since Talos should be running in maintenance mode so there are no configs yet.

As I understand Talos wants us to renew secrets each year otherwise you get locked out of the cluster. I thought this operator renewed this, did I misunderstood?

alperencelik commented:
Hi @leo-baltus, 

You have to run the operator in a Kubernetes cluster, it might be even your KinD cluster. Only thing you need to make sure is your cluster has access to your Talos Linux machines. 


> As I understand Talos wants us to renew secrets each year otherwise you get locked out of the cluster. I thought this operator renewed this, did I misunderstood?

That's true for the client certificates. Right now the operator doesn't tackle that and I think that's a great feature to implement. I just created the issue: https://github.com/alperencelik/talos-operator/issues/7. Feel free to track it there!

rothgar commented:
/m
